oprf/group/suite.go

```diff
@@ -0,0 +1,251 @@
+func (c *Ciphersuite) Order() *Scalar {
+	return &Scalar{c, c.curve.Params().N}
+}
+
+func getH2CSuite(c *Ciphersuite) (HashToPoint, error) {
+	var suite h2c.SuiteID
+
+	switch c.Name() {
+	case "OPRFP256-SHA512-ELL2-RO":
+		suite = h2c.P256_XMDSHA256_SSWU_RO_
+	case "OPRFP384-SHA512-ELL2-RO":
+		suite = h2c.P384_XMDSHA512_SSWU_RO_
+	case "OPRFP521-SHA512-ELL2-RO":
+		suite = h2c.P521_XMDSHA512_SSWU_RO_
+	default:
+		return nil, errors.New("invalid suite")
+	}
+
+	hasher, err := suite.Get(c.dst)
+	if err != nil {
+		return nil, err
+	}
+
+	return eccHasher{c, hasher}, nil
+}
+
+// HashToPoint produces a new point by hashing the input message.
+type HashToPoint interface {
+	Hash(msg []byte) (*Element, error)
+}
+
+type eccHasher struct {
+	suite  *Ciphersuite
+	hasher h2c.HashToPoint
+}
+
+func (h eccHasher) Hash(in []byte) (*Element, error) {
+	q := h.hasher.Hash(in)
+	x := q.X().Polynomial()
+	y := q.Y().Polynomial()
+
+	p := &Element{h.suite, new(big.Int), new(big.Int), true}
+	p.x.Set(x[0])
+	p.y.Set(y[0])
+
+	if !p.IsValid() {
+		return nil, errors.New("invalid point")
+	}
+
+	return p, nil
+}
+
+// HashToGroup performs a transformation to encode bytes as a Element object in the
+// group.
+func (c *Ciphersuite) HashToGroup(in []byte) (*Element, error) {
+	hasher, err := getH2CSuite(c)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := hasher.Hash(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// HashToScalar performs a transformation to encode bytes as a Scalar object in the
+// appropriate group.
+func (c *Ciphersuite) HashToScalar([]byte) (*Scalar, error) {
+	return &Scalar{}, nil
+}
+
+// RandomScalar samples a random scalar value from the field of scalars defined by the
+// group order.
+// TODO: not constant time
+func (c *Ciphersuite) RandomScalar() *Scalar {
+	N := c.Order()
+	bitLen := N.x.BitLen()
+	byteLen := (bitLen + 7) >> 3
+	buf := make([]byte, byteLen)
+
+	// rejection sampling
+	for {
+		_, err := io.ReadFull(rand.Reader, buf)
+		if err != nil {
+			panic("scalar generation failed")
+		}
+		// Mask to account for field sizes that are not a whole number of bytes.
+		var mask = []byte{0xff, 0x1, 0x3, 0x7, 0xf, 0x1f, 0x3f, 0x7f}
+		buf[0] = buf[0] & mask[bitLen%8]
+
+		// Check if scalar is in the correct range.
+		if new(big.Int).SetBytes(buf).Cmp(N.x) >= 0 {
+			continue
+		}
+		break
+	}
+
+	x := new(big.Int).SetBytes(buf)
+	return &Scalar{c, x}
+}
+
+// ScalarMultBase performs a scalar multiplication of the Generator with some scalar
+// input
+func (c *Ciphersuite) ScalarMultBase(s *Scalar) *Element {
+	gen := c.Generator()
+
+	return gen.ScalarMult(s)
+}
+
+// ByteLength returns the ByteLength of objects associated with the Ciphersuite.
+func (c *Ciphersuite) ByteLength() int {
+	return (c.curve.Params().BitSize + 7) / 8
+}
+
+// FinalizeHash computes the final hash for the suite
+func FinalizeHash(c *Ciphersuite, data, iToken, info, ctx []byte) []byte {
```

Comment:
Why is this function in `suite.go`?

Reply:
I didn't want to expose the internal hash to the outside oprf.

Reply:
But it is already exported (capital name), right? Or maybe you meant something else. I was basically asking why the function lived in suite.go rather than, say, oprf.go.

Reply:
The function is exported, but not the internal hash as part of the Ciphersuite. I put it there, so I didn't have to export the hash field.

Reply:
I'll move it  to the outside.